docs: database review 2021 english (#570)
bytebase/bytebase.com: content/blog/database-review-2021.md

|             | Funding (Round)       | Date             |
| ----------- | --------------------- | ---------------- |
| PingCAP     | $270M (D)             | 2020.11          |
| Firebolt    | $37M (A) + $127M (B)  | 2020.12 + 2021.6 |
| SingleStore | $80M (E) + $80M (F)   | 2020.12 + 2021.9 |
| CockroachDB | $160M (E) + $278M (F) | 2021.1 + 2021.12 |
| Yugabyte    | $48M (B) + $188M (C)  | 2021.3 + 2021.10 |
| PlanetScale | $30M (B) + $50M (C)   | 2021.6 + 2021.10 |
| ClickHouse  | $50M (A) + $250M (B)  | 2021.9 + 2021.10 |

The AI hype is cooling down, the blockchain opportunists jump from Bitcoin to the metaverse, then people read Snowflake's and MongoDB's earnings release and realize what makes the real money, the classic OLTP / OLAP (Online Transaction Processing and Analysis) database that every company relies on every day.

## Proven Technology

There has been underlying technology breakthrough in the database industry in these two years. After a round of NoSQL fuss 10 years ago, we have returned to the relational model with SQL as the interface. The only notable addition is the distributed architecture inspired by Google Spanner.

> Although the concept of autonomous/self-driving DB has been around for many years, the database still needs DBAsit. Query and Index Optimization, the most suitable scenario for AI, has not been applied in the mainstream database since [Google's 2018 paper](https://research.google/pubs/pub46518/).

Author: I coined from babysit

ClickHouse has existed for some time, incubated in the Russian Internet giant Yandex, ByteDance, the owner of TikTok also claims to have a ClickHouse cluster with over 15,000 nodes. 2021 is the year of ClickHouse, in August, the just formed ClickHouse Inc, immediately announced two funding rounds totaling $300 millions.

If compared with Snowflake, ClickHouse documentation, ecosystem tools are pretty behind. But ClickHouse has unmatched single-table query performance and also supports on premise deployment, it fits the fast growing internet business requirements perfectly.

So if the technology is really strong enough, it is possible to put the product aside, for a moment. Still ClickHouse has formed ClickHouse Inc to build the cloud product. It already seems to be an afterthought after [Gigapipe](https://gigapipe.com/) and [Firebolt](https://firebolt.io) debuts.

### Firebolt

If Firebolt is 50% of what it advertises, it's time to short Snowflake.

Let’s take a look at the Firebolt manifesto:

![_](/static/blog/database-review-2021-bytebase/firebolt-leap.webp)
![_](/static/blog/database-review-2021-bytebase/snowflake-firebolt.webp)

> Of course Snowflake is a great product, and over the years it has grown into a one-stop platform that Firebolt, as a startup, is nowhere near in terms of breadth of product. But Firebolt is targeting Snowflake's core competence Price / Performance / Usability, and their website claims:

Reviewer: ```suggestion
Of course, Snowflake is a great product, and over the years it has grown into a one-stop platform that Firebolt, as a startup, is nowhere near in terms of product breadth. But Firebolt is targeting Snowflake's core competence Price / Performance / Usability, and their website claims:
```

In fact, PG has won the DB-Engine DBMS of the year award 3 times in the last 5 years (Snowflake / MySQL in 2021 and 2019 respectively). PG is gaining popularity for several reasons (since PG's main competitor is MySQL, the following is more of a comparison with MySQL):

1. Licensing concern after Oracle acquires MySQL. Personally, I think Oracle has made a lot of important improvements to MySQL after the acquisition, but licensing is a real risk, especially if the other side is Oracle.
2. PG has the inherent architecture advantage, its advanced/extended features are much ahead of MySQL. For example, although MySQL also has Geospatial, JSON functionality, it is still quite inferior to PG's PostGIS, JSON-B.
3. PG has a much more sophisticated query engine than MySQL. Not long ago, a former MySQL query engine engineer [ranted](https://blog.sesse.net/blog/tech/2021-12-05-16-41_leaving_mysql.html) before leaving the MySQL team. In fact, PG can have above average AP capabilities with a little tuning (many OLAP products like AWS Redshift are based on PostgreSQL).

> 4. The logical isolation of PG is at the database level, where a PG instance can create many databases and then assign each database to a different tenant. Although MySQL instances can also create multiple databases, because the databases are connected by default, this makes it necessary to assign a separate MySQL instance to each tenant. In single-tenant mode, MySQL's ease of use is superior, but in today's SaaS, multi-tenant service scenario, PG has a big advantage. That's why PG has been adopted from the earliest Heroku, to the new Heroku-like render, and Supabase. They can offer a low or even free database plan because they can serve many users with a single PG instance.

Reviewer: ```suggestion
4. The logical isolation of PG is at the database level, where a PG instance can create many databases and then assign each database to a different tenant. Although MySQL instances can also create multiple databases, because the databases are connected by default, this makes it necessary to assign a separate MySQL instance to each tenant. In single-tenant mode, MySQL's ease of use is superior, but in today's SaaS, multi-tenant service scenario, PG has a big advantage. That's why PG has been adopted by Heroku early on, to the new Heroku-like render, and Supabase. They can offer a low or even free database plan because they can serve many users with a single PG instance.
```

PS is batteries included. When a company chooses PG, it gets a database with OLTP, OLAP, Document (JSON-B), FTS (tsvector/tsquery), Time-series (TimescaleDB), Geospatial (PostGIS), Multi-tenancy capabilities (batteries included). To summarize it in a simple formula:

PostgreSQL = MySQL + Poor man version of (ClickHouse + MongoDB\* + Elasticsearch + InfluxDB) + Geospatial + Multi-tenancy

> - MongoDB 有分布式能力，这点比 PG 要强，但是论单机 KV 能力的话，PG 不少点其实更强。

Reviewer: ```suggestion
- MongoDB has distributed capability, which is stronger than PostgreSQL, but in terms of standalone KV capability, PostgreSQL is actually stronger in quite a few points.
```

> - Geospatial 在开源届公认的方案就是 PostGIS，所以如果业务上地理信息处理是核心，那 PostgreSQL 其实也是唯一的选择。

Reviewer: ```suggestion
- In the open-source world, PostGIS is the accepted Geospatial solution, so if geographic info processing is the core of the business, PostgreSQL is actually the only choice.
```

> - Multi-tenancy 这里指的更多是应用逻辑上的隔离，但还做不到像 Google Spanner 那样真正租户级别的物理隔离。

Reviewer: ```suggestion
- Multi-tenancy here refers more to logical application isolation, but does not yet allow for true tenant-level physical isolation like Google Spanner.
```

> 
> PostgreSQL = MySQL + Poor man version of (ClickHouse + MongoDB\* + Elasticsearch + InfluxDB) + Geospatial + Multi-tenancy
> 
> - MongoDB 有分布式能力，这点比 PG 要强，但是论单机 KV 能力的话，PG 不少点其实更强。
> 
> - Geospatial 在开源届公认的方案就是 PostGIS，所以如果业务上地理信息处理是核心，那 PostgreSQL 其实也是唯一的选择。
> 
> - Multi-tenancy 这里指的更多是应用逻辑上的隔离，但还做不到像 Google Spanner 那样真正租户级别的物理隔离。
> 
> Although PG can't compete with the ClickHouse in single-table query performance, or Elasticsearch in text retrieval, and even in transactional processing capabilities, it’s also [inferior to MySQL](https://eng.uber.com/postgres-to-mysql-migration), but the premise of having a single database system to serve all scenarios is too attractive.

Author: ```suggestion

PostgreSQL = MySQL + Poor man version of (ClickHouse + MongoDB\* + Elasticsearch + InfluxDB) + Geospatial + Multi-tenancy

Although PG can't compete with the ClickHouse in single-table query performance, or Elasticsearch in text retrieval, and even in transactional processing capabilities, it’s also [inferior to MySQL](https://eng.uber.com/postgres-to-mysql-migration), but the premise of having a single database system to serve all scenarios is too attractive.
```

![_](/static/blog/database-review-2021-bytebase/planetscale-workflow.webp)

In 2021, two companies that have re-defined the developer workflow today went IPO, GitLab and HashiCorp, both with a market cap of over $10 billion. PlanetScale folks also see this opportunity in the database market and they have also partnered with Vercel, the company about to redefine the frontend developer workflow.

![_](/static/blog/database-review-2021-bytebase/vercel.webp)

The tech stack started with LAMP (Linux + Apache + MySQL + PHP), then came MEAN (MongoDB + Express + Angular + Node), and now there is a new VP (Vercel + PlanetScale) stack that could introduce another paradigm shift.

![_](/static/blog/database-review-2021-bytebase/planetscale.webp)

And when it comes to acronym names, there's another one coming up in 2021

![_](/static/blog/database-review-2021-bytebase/gosh-stack.webp)

Oh my **GoSH**.

### SQLite

> While the authors of curl [may not agree](https://daniel.haxx.se/blog/2021/10/21/the-most-used-software-components-in-the-world/) that SQLite is the most deployed software component in the world, SQLite is undoubtedly the most deployed database, with perhaps dozens or hundreds on a single phone, and in almost every app from Notes to WhatsApp. SQLite is not a complete database system (DBMS), but it has a complete storage engine and SQL interface, and it has goode enough performance for most application needs. It's a library and use a single file to hold data, which makes application access and manage data very easy. In 2021, several interesting projects have emerged from the SQLite ecosystem:

Reviewer: ```suggestion
While the authors of curl [may not agree](https://daniel.haxx.se/blog/2021/10/21/the-most-used-software-components-in-the-world/) that SQLite is the most deployed software component in the world, SQLite is undoubtedly the most deployed database, with perhaps dozens or hundreds on a single phone, and in almost every app from Notes to WhatsApp. SQLite is not a complete database system (DBMS), but it has a complete storage engine and SQL interface, and it has good enough performance for most application needs. It's a library and uses a single file to hold data, which makes application access and data management very easy. In 2021, several interesting projects have emerged from the SQLite ecosystem:
```

1. PlanetScale will be widely adopted, Vercel + PlanetScale, the VP stack will bring a new paradigm shift to the developer workflow, especially to attract a lot of front-end and full-stack developers.
2. There will be a new database coming out, and the main selling point will be developer workflow.
3. There will be startups building ClickHouse tools, and receive a lot of funding.
4. Firebolt will be the fastest-funded database company ever.
5. There will be killer solution based on SQLite.
6. There will be an open source tool built for application database development 😀

## Parting words

1. The database itself is just a technology. Technology can’t sell, only solution does. So what needs to be built is the cloud platform, tooling, developer ecosystem and best practices around the database development workflow.
2. The database engine is the core, but performance is only one of the puzzle, the engine space is already crowded, but there are plenty rooms around the database development workflow. To differentiate, investing developer experience and betting on defining the new developer workflow.
3. Pay attention to SQLite.
4. Choose PostgreSQL, you won’t regret it.

> See you in 2023!

Reviewer: but this article says 2021 👀 where is 2022 tho

Author: The article is written in 2022, so if we write the next article for 2022, then it will be written in 2023.
